Error updates vol. 3

chmeliik opened the pull request:
Main changes:

* re-raise input validation errors as InvalidInput
* return rc=2 for invalid usage
* log errors before printing them (https://github.com/containerbuildsystem/cachi2/pull/20#issuecomment-1308881540)

WIP: how to log multiline errors

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] New code has type annotations
- [ ] Docs updated (if applicable)
- [ ] [Issue](https://github.com/containerbuildsystem/cachi2/issues/13) that tracks breaking changes to the OSBS2 integration is updated (if applicable)

chmeliik commented on cachi2/interface/cli.py:
```diff
@@ -204,11 +195,14 @@ def combine_flags() -> list[str]:
-        output_dir=output,
-        packages=parsed_packages,
-        flags=combine_flags(),
+    request = parse_user_input(
+        Request.parse_obj,
+        {
+            "source_dir": source,
+            "output_dir": output,
+            "packages": parsed_packages,
+            "flags": combine_flags(),
+        },
```
I'm not entirely happy about having to call the `parse_user_input` function to get InvalidInput errors. But it was the best I could come up with, alternatives being:

* make a BaseInput class and override `parse_obj`, `parse_file` and `parse_raw`
  * I didn't tests this one, didn't like it because you could still `__init__` directly and get the raw pydantic error
* ~~make a BaseInput class and override `__init__`~~
  * breaks grouped error reporting (see below)
* override `Request.__init__`
  * works, but why is Request the only input model that gets special treatment?

grouped error reporting:
```
Error: InvalidInput: 2 validation errors for user input
packages -> 0 -> type
  unexpected value; permitted: 'gomod' (given=foo; permitted=('gomod',))
packages -> 1 -> type
  unexpected value; permitted: 'gomod' (given=bar; permitted=('gomod',))
``` 
but if the Package model raises InvalidInput instead of pydantic.ValidationError:
```
Error: InvalidInput: 1 validation error for user input
type
  unexpected value; permitted: 'gomod' (given=foo; permitted=('gomod',))
```

brunoapimentel replied:
Can't we treat this in the CLI `handle_error` decorator level?  You would also catch `pydantic.ValidationError` there, and reraise it as InvalidInput.

chmeliik replied:
The thing is `pydantic.ValidationError` is not always invalid input. We also have output models, and invalid output would usually be caused by programmer error. But if we're confident we won't mess it up, maybe that's OK :smile:

brunoapimentel replied:
Not sure if that's much better, but we can do this in the `handle_error` decorator:

```
except pydantic.ValidationError as e:
   if e.model.__module__ == "cachi2.core.models.input":
      msg, details = _present_user_input_error(e)
      raise InvalidInput(f"{msg}\n{details}") from e
   else:
       raise e
```

chmeliik replied:
hmm, that feels a little too python-hacky to me

chmeliik replied:
I tried that out and eventually reached a solution that felt reasonable.

Now, we override `_BaseInput.__init__` and raise an error that inherits from _both_ InvalidInput and pydantic.ValidationError. That way, pydantic still groups together sub-model validation errors as expected.

Because I still wanted the error to show up as InvalidInput, I came up with the idea of private error classes which report the type name of a non-private parent class.

Let me know if that seems acceptable or too magical :smile:

lkolacek replied:
It seems magical to me, but I believe I get it. Will see for how long I'll remember it. :) If the approach in the 6th commit will be the one that is used are you going to place code into the 1st commit and drop it?

chmeliik replied:
:+1: yeah putting it in the 1st commit would be better

chmeliik replied:
But yeah, if it's too much magic maybe I'll just keep the original approach with having to call a specific function when parsing user input

chmeliik commented on cachi2/interface/cli.py:
```diff
@@ -102,6 +102,7 @@ def fetch_deps(
         ...,  # Ellipsis makes this option required
         help="Specify package (within the source repo) to process. See usage examples.",
         metavar="PKG",
+        callback=lambda args: [_if_json_then_validate(arg) for arg in args],
```
Note: if the validation error happens in the callback, our error handler will **not** log it. This is true for all default Typer errors, such as non-existent parameters, missing `--source` directory etc.

How concerned are we about that? (Personally, not that much)

brunoapimentel replied:
I looks like all other Typer errors, I'd say it's fine.

chmeliik commented:
PTAL, this is mostly finished apart from multiline error logging (and the unit tests for error logging)

For ^ I present two options: 
* no multine error: https://github.com/containerbuildsystem/cachi2/pull/22/commits/cbfa24155ae0cf149c076588242e5c7a0ae18803
* replace newlines with literal `\n`: https://github.com/containerbuildsystem/cachi2/pull/22/commits/20932fd9d6e8961b5e22d34dfe070b8bb9569e2a

Aesthetically I prefer the first one, but the second is probably more practical

brunoapimentel commented:
> PTAL, this is mostly finished apart from multiline error logging (and the unit tests for error logging)
> 
> For ^ I present two options:
> 
> * no multiline errors: [cbfa241](https://github.com/containerbuildsystem/cachi2/commit/cbfa24155ae0cf149c076588242e5c7a0ae18803)
> * replace newlines with literal `\n`: [20932fd](https://github.com/containerbuildsystem/cachi2/commit/20932fd9d6e8961b5e22d34dfe070b8bb9569e2a)
> 
> Aesthetically I prefer the first one, but the second is probably more practical
I'm thinking we could settle for approach 1 for now, but there's a few future consequences we could stumble upon:
- New errors that are multi-line would need to be treated individually as they're added
- When using Splunk, if the single line message is not enough, we'd need to go to the raw output to find the rest of the error log
- If a user configures the logger to write to a file, the stderr output would not be present there

chmeliik commented:
> When using Splunk, if the single line message is not enough, we'd need to go to the raw output to find the rest of the error log

Yeah that's the main one, I'd maybe rather go with practicality here and log the whole thing

chmeliik commented:
Added one unit test check for error logging, removed WIP commits (now logging multiline error messages on one line, with a literal `\n`)

Should be done-ish
